3064 -Reparse Meta Model

elipe17 opened the pull request:
## Summary of Changes
- Created new meta model for reparse executions
- Added logic to enforce sequential reparse execution
- Added timeout logic to reparse execution as a failsafe
- Moved datafiles admin config to separate directory
- Added reparse event filtering on datafiles admin page
- Removed Elastic options from reparse command. `-a` enforces new indices and that is the only time they are recreated.
- Updated celery log level and forced single task execution to preserve memory
- Converted "re-parse" to "reparse" to unify language
Pull request closes #3064 

## How to Test
```
cd tdrs-frontend && docker-compose up
cd tdrs-backend && docker-compose up --build
```

1. Open http://localhost:3000/ and sign in.
2. Submit some files
3. execute reparse command and try to break it

## Deliverables
_More details on how deliverables herein are assessed included [here](https://github.com/raft-tech/TANF-app/blob/develop/docs/How-We-Work/our-priorities-values-expectations.md#Deliverables)._

### [Deliverable 1: Accepted Features](https://github.com/raft-tech/TANF-app/blob/develop/docs/How-We-Work/our-priorities-values-expectations.md#Deliverable-1-Accepted-Features)

Checklist of ACs:
- [x] Meta data model is created
- [x] Meta data model is serialized to the DB per run of the `clean_and_reparse` command
- [x] Migration is named appropriately
- [x] Re-parse commands are guarenteed to run sequentially
- [x] Re-parse model "re-parse completed" datetime field has the correct datetime logic
- [x] Testing Checklist has been run and all tests pass
- [x] README is updated, if necessary

### [Deliverable 2: Tested Code](https://github.com/raft-tech/TANF-app/blob/develop/docs/How-We-Work/our-priorities-values-expectations.md#Deliverable-2-Tested-Code)

+ Are all areas of code introduced in this PR meaningfully tested?
  + [ ] If this PR introduces backend code changes, are they meaningfully tested?
  + [ ] If this PR introduces frontend code changes, are they meaningfully tested?
+ Are code coverage minimums met?
  + [x] Frontend coverage: [_insert coverage %_] (see `CodeCov Report` comment in PR)
  + [x] Backend coverage: [_insert coverage %_] (see `CodeCov Report` comment in PR)

### [Deliverable 3: Properly Styled Code](https://github.com/raft-tech/TANF-app/blob/develop/docs/How-We-Work/our-priorities-values-expectations.md#Deliverable-3-Properly-Styled-Code)

+ [x] Are backend code style checks passing on CircleCI?
+ [x] Are frontend code style checks passing on CircleCI?
+ [x] Are code maintainability principles being followed?

### [Deliverable 4: Accessible](https://github.com/raft-tech/TANF-app/blob/develop/docs/How-We-Work/our-priorities-values-expectations.md#Deliverable-4-Accessibility)

+ [ ] Does this PR complete the epic? 
+ [ ] Are links included to any other gov-approved PRs associated with epic?
+ [ ] Does PR include documentation for Raft's a11y review? 
+ [ ] Did automated and manual testing with `iamjolly` and `ttran-hub` using Accessibility Insights reveal any errors introduced in this PR?


### [Deliverable 5: Deployed](https://github.com/raft-tech/TANF-app/blob/develop/docs/How-We-Work/our-priorities-values-expectations.md#Deliverable-5-Deployed)

+ [x] Was the code successfully deployed via automated CircleCI process to development on Cloud.gov?

### [Deliverable 6: Documented](https://github.com/raft-tech/TANF-app/blob/develop/docs/How-We-Work/our-priorities-values-expectations.md#Deliverable-6-Code-documentation)

+ [x] Does this PR provide background for why coding decisions were made?
+ [ ] If this PR introduces backend code, is that code easy to understand and sufficiently documented, both inline and overall?
+ [ ] If this PR introduces frontend code, is that code easy to understand and sufficiently documented, both inline and overall?
+ [ ] If this PR introduces dependencies, are their licenses documented?
+ [x] Can reviewer explain and take ownership of these elements presented in this code review?

### [Deliverable 7: Secure](https://github.com/raft-tech/TANF-app/blob/develop/docs/How-We-Work/our-priorities-values-expectations.md#Deliverable-7-Secure)

+ [x] Does the OWASP Scan pass on CircleCI?
+ [ ] Do manual code review and manual testing detect any new security issues?
+ [ ] If new issues detected, is investigation and/or remediation plan documented? 

### [Deliverable 8: User Research](https://github.com/raft-tech/TANF-app/blob/develop/docs/How-We-Work/our-priorities-values-expectations.md#Deliverable-8-User-Research)

Research product(s) clearly articulate(s):
+ [ ] the purpose of the research
+ [ ] methods used to conduct the research 
+ [ ] who participated in the research
+ [ ] what was tested and how
+ [ ] impact of research on TDP
+ [ ] (_if applicable_) final design mockups produced for TDP development

elipe17 committed:
- Add reparse meta model
- Add sequential execution logic
- Add parse logic for updating meta model
- make migration for meta model

elipe17 committed:
- Use string reference for datafile meta model to avoid circular import
- Update util

elipe17 committed:
- Pass model into functions instead of pk

elipe17 committed:
- rename to reparse to keep consistent
- add admin model

elipe17 committed:
- update migration content

elipe17 committed:
- Kill celery after every task execution to relieve memory pressure

elipe17 committed:
- Remove invalid checks

elipe17 committed:
- remove print

elipe17 committed:
- Added a status field to meta model
- Added convenient function to assert completion status

elipe17 committed:
- add rpv to index name

elipe17 committed:
- lint

elipe17 committed:
- updated permisions in test

elipe17 committed:
- Added timeout field to model
- Added a timeout algorithm based on local testing
- Added logging to capture when we are reparsing after exceeding a timeout

elipe17 committed:
- Updated datafile relation to be a many to many relation to allow tracking of reparse events
- Updated meta model class to support many to many relation for helper funcs

elipe17 committed:
- update reparse to not delete datafile to preserve the many to many relationship

elipe17 committed:
- lint

elipe17 committed:
- Undo debug adds

elipe17 committed:
- remove invalid comment

jtimpe commented on tdrs-backend/tdpservice/search_indexes/management/commands/clean_and_reparse.py:
```diff
@@ -156,8 +179,53 @@ def __handle_datafiles(self, files, log_context):
+
+    def __calculate_timeout(self, num_files, num_records):
+        """Estimate a timeout parameter based on the number of files and the number of records."""
+        MEDIAN_PARSE_TIME_ON_LOCAL = 0.0005574226379394531
```
parameterize

elipe17 replied:
Done in 3065 PR

elipe17 committed:
- remove sequential checks

elipe17 committed:
- parametrize line parse time

elipe17 committed:
- Add admin resources for the many to many relationship
- Add helper function to update datafile state during parsing

elipe17 committed:
Merge branch '3064-reparse-meta' of https://github.com/raft-tech/TANF-app into 3065-sequential-reparse

elipe17 committed:
- resolve conflicts

elipe17 committed:
- lint

elipe17 committed:
Merge branch '3064-reparse-meta' of https://github.com/raft-tech/TANF-app into 3065-sequential-reparse

jtimpe commented:
I get a `duplicate key error` and lose all record data when trying to replicate the "double button click" (sequential execution) for which you removed the handling. just want to highlight the risk there
```bash
[2024-08-09 14:22:13,973: ERROR/ForkPoolWorker-12] Encountered Database exception in parser_task.py: 
web-1           | duplicate key value violates unique constraint "parsers_datafilesummary_datafile_id_880a2f4d_uniq"
web-1           | DETAIL:  Key (datafile_id)=(4) already exists.
```
<img width="349" alt="image" src="https://github.com/user-attachments/assets/3d577e36-d5da-42cd-8c46-ab80332886d6">

jtimpe reviewed (approved):
Works great!

elipe17 committed:
- Remove RPM links

raftmsohani commented:
@elipe17 I am still not clear why we need many to many relationship, and I would like to avoid them if possible. There are many things that can go wrong with them, and leave junk data in the DB.
Many if you could elaborate more on why many-to-many is needed I can be convinced!

raftmsohani reviewed (commented):
Have question about many to many relationship as I think reparse_meta_models can hold one to many relation with datafile

elipe17 committed:
Merge branch 'develop' into 3064-reparse-meta

elipe17 committed:
Merge branch 'develop' into 3064-reparse-meta

elipe17 committed:
Merge branch 'develop' into 3064-reparse-meta

andrew-jameson commented on tdrs-backend/tdpservice/search_indexes/admin/reparse_meta.py:
```diff
@@ -0,0 +1,20 @@
+        'db_backup_location',
+    ]
+
+    list_filter = [
+        'success',
+        'finished'
```
Can we add these as filters?

```
 # Options used to select the files to reparse
    fiscal_quarter = models.CharField(max_length=2, null=True)
    fiscal_year = models.PositiveIntegerField(null=True)
 ```

elipe17 committed:
Merge branch 'develop' into 3064-reparse-meta

elipe17 committed:
- added fiscal year/quarter filters

elipe17 committed:
Merge branch '3064-reparse-meta' of https://github.com/raft-tech/TANF-app into 3065-sequential-reparse

andrew-jameson commented:
Both locally and on a11y, I am not getting a finished reparse run. I initially just used `-a` to gather the dozen or so files i had uploaded then tried breaking it up by year, same results. I waited for logs to indicate no more parsing was happening before initiating the next run. Used `task clean` prior to building so its completely fresh. a11y also has a fresh DB due to some issues. Will retry against raft env.

<details><summary>Screenshots</summary>
![Screenshot 2024-08-15 at 11 44 34 AM](https://github.com/user-attachments/assets/11b3ad8f-cafc-4948-bb47-9eb7e0ce83e8)

![Screenshot 2024-08-15 at 11 56 28 AM](https://github.com/user-attachments/assets/2abdba46-0516-42fe-8b63-0c96f8c1fe22)
![Screenshot 2024-08-15 at 11 56 43 AM](https://github.com/user-attachments/assets/ce449fa2-6736-4896-8626-c99af8633eb2)
</details>


```
2024-08-15 11:55:21 [2024-08-15 15:55:21,346: INFO/ForkPoolWorker-52] DataFile parsing started for file ADS.E2J.NDM1.TS01
2024-08-15 11:55:21 2024-08-15 15:55:21,426 DEBUG fields.py::parse_value:L47 :  Field: 'tribe_code' at position: [14, 17) is empty.
2024-08-15 11:55:21 2024-08-15 15:55:21,426 DEBUG fields.py::parse_value:L47 :  Field: 'tribe_code' at position: [14, 17) is empty.
2024-08-15 11:55:21 2024-08-15 15:55:21,426 DEBUG parse.py::parse_datafile:L46 :  Datafile has encrypted fields: True.
2024-08-15 11:55:21 2024-08-15 15:55:21,426 DEBUG parse.py::parse_datafile:L47 :  Datafile: {id: 21, filename: ADS.E2J.FTP1.TS06, STT: Alabama (01), S3 location: data_files/2023/Q1/1/Active Case Data/ADS.E2J.FTP1.TS06}, is Tribal: False.
2024-08-15 11:55:21 2024-08-15 15:55:21,426 DEBUG parse.py::parse_datafile:L51 :  Program type: TAN, Section: A.
2024-08-15 11:55:21 2024-08-15 15:55:21,427 INFO parse.py::parse_datafile:L95 :  Preparser Error -> Rpt Month Year is not valid: Submitted reporting year:2020, quarter:Q4 doesn't match file reporting year:2023, quarter:Q1.
2024-08-15 11:55:21 2024-08-15 15:55:21,427 DEBUG parse.py::bulk_create_errors:L155 :  Bulk creating ParserErrors.
2024-08-15 11:55:21 2024-08-15 15:55:21,429 INFO parse.py::bulk_create_errors:L158 :  Created 1/1 ParserErrors.
2024-08-15 11:55:21 2024-08-15 15:55:21,439 INFO parser_task.py::parse:L41 :  Parsing finished for file -> {id: 21, filename: ADS.E2J.FTP1.TS06, STT: Alabama (01), S3 location: data_files/2023/Q1/1/Active Case Data/ADS.E2J.FTP1.TS06} with status Rejected and 1 errors.
2024-08-15 11:55:21 [2024-08-15 15:55:21,439: INFO/ForkPoolWorker-52] Parsing finished for file -> {id: 21, filename: ADS.E2J.FTP1.TS06, STT: Alabama (01), S3 location: data_files/2023/Q1/1/Active Case Data/ADS.E2J.FTP1.TS06} with status Rejected and 1 errors.
2024-08-15 11:55:21 [2024-08-15 15:55:21,441: INFO/ForkPoolWorker-52] Task tdpservice.scheduling.parser_task.parse[22642997-1f01-4f61-bc74-0cef780d0247] succeeded in 0.10376212500000292s: None
```

andrew-jameson commented on tdrs-backend/tdpservice/search_indexes/management/commands/clean_and_reparse.py:
```diff
@@ -20,16 +20,16 @@
+        parser.add_argument("-a", "--all", action='store_true', help="Clean and reparse all datafiles. If selected, "
                             "fiscal_year/quarter aren't necessary.")
-        parser.add_argument("-n", "--new_indices", action='store_true', help="Move re-parsed data to new Elastic "
+        parser.add_argument("-n", "--new_indices", action='store_true', help="Move reparsed data to new Elastic "
                             "indices.")
         parser.add_argument("-d", "--delete_indices", action='store_true', help="Requires new_indices. Delete the "
                             "current Elastic indices.")
```
Not having these (`-n` and `-d`) selected results in failures and inconsistency, we should remove them as parameters and just always have these executed as true, as far as my understanding.

elipe17 replied:
Can you digress on the failures you saw? In my opinion, the only time we should specify these options is when we want to reindex the entire DB and get elastic consistent. I think having elastic indexes dedicated to a subset of the data does not make sense. In the future they only want to reparse single files. In that case we would create 21 new indexes to store a single file. We will overload elastic very quickly doing that. With that said, I would be open to removing the availability of the options in favor of forcing new indices if and only if `-a` is specified.

andrew-jameson replied:
I no longer have the errors in my shell scrollback, but it was to the effect that running -a on different environments was causing drift between ES and postgres. Attempting to recreate this behavior with the deployed code from #3136 no longer shows the behavior as you alluded to [here](https://github.com/raft-tech/TANF-app/pull/3126#issuecomment-2296553787). 

Given the discrepancies in behavior, we should merge that PR into this one for simplicity. Yes, let's remove the availability for those when using -a and force new indices. After this change, I'll be good to send this over to QASP.

andrew-jameson commented on tdrs-backend/tdpservice/search_indexes/management/commands/clean_and_reparse.py:
```diff
@@ -38,22 +38,22 @@ def __get_log_context(self, system_user):
+            log("Database backup FAILED. Clean and reparse NOT executed. Database and Elastic are CONSISTENT!",
                 logger_context=log_context,
                 level='error')
             raise e
```
in the below function call of `__handle_elastic`, we should always set delete and new indices to happen given my above comment.

andrew-jameson reviewed (changes requested):
Comments about indices flags being needed on runs.

elipe17 commented:
> Both locally and on a11y, I am not getting a finished reparse run. I initially just used `-a` to gather the dozen or so files i had uploaded then tried breaking it up by year, same results. I waited for logs to indicate no more parsing was happening before initiating the next run. Used `task clean` prior to building so its completely fresh. a11y also has a fresh DB due to some issues. Will retry against raft env.
> 
> Screenshots
> ```
> 2024-08-15 11:55:21 [2024-08-15 15:55:21,346: INFO/ForkPoolWorker-52] DataFile parsing started for file ADS.E2J.NDM1.TS01
> 2024-08-15 11:55:21 2024-08-15 15:55:21,426 DEBUG fields.py::parse_value:L47 :  Field: 'tribe_code' at position: [14, 17) is empty.
> 2024-08-15 11:55:21 2024-08-15 15:55:21,426 DEBUG fields.py::parse_value:L47 :  Field: 'tribe_code' at position: [14, 17) is empty.
> 2024-08-15 11:55:21 2024-08-15 15:55:21,426 DEBUG parse.py::parse_datafile:L46 :  Datafile has encrypted fields: True.
> 2024-08-15 11:55:21 2024-08-15 15:55:21,426 DEBUG parse.py::parse_datafile:L47 :  Datafile: {id: 21, filename: ADS.E2J.FTP1.TS06, STT: Alabama (01), S3 location: data_files/2023/Q1/1/Active Case Data/ADS.E2J.FTP1.TS06}, is Tribal: False.
> 2024-08-15 11:55:21 2024-08-15 15:55:21,426 DEBUG parse.py::parse_datafile:L51 :  Program type: TAN, Section: A.
> 2024-08-15 11:55:21 2024-08-15 15:55:21,427 INFO parse.py::parse_datafile:L95 :  Preparser Error -> Rpt Month Year is not valid: Submitted reporting year:2020, quarter:Q4 doesn't match file reporting year:2023, quarter:Q1.
> 2024-08-15 11:55:21 2024-08-15 15:55:21,427 DEBUG parse.py::bulk_create_errors:L155 :  Bulk creating ParserErrors.
> 2024-08-15 11:55:21 2024-08-15 15:55:21,429 INFO parse.py::bulk_create_errors:L158 :  Created 1/1 ParserErrors.
> 2024-08-15 11:55:21 2024-08-15 15:55:21,439 INFO parser_task.py::parse:L41 :  Parsing finished for file -> {id: 21, filename: ADS.E2J.FTP1.TS06, STT: Alabama (01), S3 location: data_files/2023/Q1/1/Active Case Data/ADS.E2J.FTP1.TS06} with status Rejected and 1 errors.
> 2024-08-15 11:55:21 [2024-08-15 15:55:21,439: INFO/ForkPoolWorker-52] Parsing finished for file -> {id: 21, filename: ADS.E2J.FTP1.TS06, STT: Alabama (01), S3 location: data_files/2023/Q1/1/Active Case Data/ADS.E2J.FTP1.TS06} with status Rejected and 1 errors.
> 2024-08-15 11:55:21 [2024-08-15 15:55:21,441: INFO/ForkPoolWorker-52] Task tdpservice.scheduling.parser_task.parse[22642997-1f01-4f61-bc74-0cef780d0247] succeeded in 0.10376212500000292s: None
> ```

@andrew-jameson the code that handles tracking failed files (think S3 exception we don't catch) or files that exit parsing early due to cat1 errors is in the follow on PR since it is required for sequential execution and not general metadata tracking.

elipe17 committed:
Merge branch 'develop' into 3064-reparse-meta

elipe17 committed:
- add timeout None check in assertion

andrew-jameson commented on tdrs-backend/tdpservice/data_files/models.py:
```diff
@@ -152,6 +152,11 @@ class Meta:
 
+    reparse_meta_models = models.ManyToManyField("search_indexes.ReparseMeta",
+                                                 help_text="Reparse events this file has been associated with.",
+                                                 related_name="reparse_meta_models"
```
```suggestion
                                                 related_name="datafiles"
```

andrew-jameson commented:
Usability change for sysadmins and developers: Data Files page filter by a ReparseMeta model object. 

for my own notes:
```python
meta7 = ReparseMeta.objects.get(id=7)
datafiles = DataFile.objects.filter(reparse_meta_models=meta7)
# equivalent is ~ meta7.reparse_meta_models.all()
[print("{}:{}".format(d.stt,d.fiscal_year)) for d in datafiles]

## files associated with 
Arkansas (05):2024 - Q1 (Oct - Dec)
Arkansas (05):2024 - Q1 (Oct - Dec)
Alabama (01):2024 - Q1 (Oct - Dec)
Chippewa-Cree Indians of the Rocky Boy's Reservation (043):2024 - Q1 (Oct - Dec)
Chippewa-Cree Indians of the Rocky Boy's Reservation (043):2024 - Q1 (Oct - Dec)
Chippewa-Cree Indians of the Rocky Boy's Reservation (043):2024 - Q1 (Oct - Dec)
Chippewa-Cree Indians of the Rocky Boy's Reservation (043):2024 - Q1 (Oct - Dec)
Florida (12):2024 - Q1 (Oct - Dec)
Florida (12):2024 - Q1 (Oct - Dec)
Alabama (01):2024 - Q1 (Oct - Dec)
Arkansas (05):2024 - Q1 (Oct - Dec)
Alabama (01):2024 - Q1 (Oct - Dec)
```

andrew-jameson reviewed (changes requested):
0. merge 3065 into this
1. Usability change for sysadmins and developers: Data Files page filter by a ReparseMeta model object.
2. update related names on datafile manyToMany relation
3. force new indices when using -a flag

elipe17 committed:
Merge pull request #3136 from raft-tech/3065-sequential-reparse

3065 -Sequential Reparse

elipe17 committed:
- Create admin dir for datafiles
- Add filter for reparse events on df page
- remove `-n` and `-d` options from command
- update relation name

elipe17 committed:
- fix lint

elipe17 committed:
- fix import

elipe17 commented on tdrs-backend/tdpservice/data_files/admin/filters.py:
```diff
@@ -0,0 +1,59 @@
+
+    def queryset(self, request, queryset):
+        """Sort queryset to show datafiles associated to the most recent reparse event."""
+        if self.value() is not None and queryset.exists():
```
Ask Alex if we want to filter on the actual ID of the meta model as opposed to the "latest" meta model.

andrew-jameson reviewed (approved):
Lets go 🚀

ADPennington commented:
per standup today #3064 and #3065 work reflected in this PR. I started testing this morning.

ADPennington commented on docs/Technical-Documentation/clean-and-reparse.md:
```diff
@@ -201,22 +201,22 @@ space:          tanf-dev
-  -n, --new_indices     Move re-parsed data to new Elastic indices.
+                        Reparse all files in the fiscal year, e.g. 2021.
+  -a, --all             Clean and reparse all datafiles. If selected, fiscal_year/quarter aren't necessary.
+  -n, --new_indices     Move reparsed data to new Elastic indices.
```
@andrew-jameson @jtimpe @elipe17 @raftmsohani can this portion of the documentation be updated to remove the `-n` and `-d` elements? 

```shell=
vcap@...:~$ python manage.py clean_and_reparse -y 2022 -q Q1 -d -n
usage: manage.py clean_and_reparse [-h] [-q {Q1,Q2,Q3,Q4}]
                                   [-y FISCAL_YEAR] [-a]
                                   [--configuration CONFIGURATION]
                                   [--version] [-v {0,1,2,3}]
                                   [--settings SETTINGS]
                                   [--pythonpath PYTHONPATH]
                                   [--traceback] [--no-color]
                                   [--force-color] [--skip-checks]
manage.py clean_and_reparse: error: unrecognized arguments: -d -n

```

ADPennington commented:
@elipe17 @andrew-jameson @jtimpe @raftmsohani I'm currently `blocked` on testing this PR in qasp environment. I attempted to reparse this morning for FY2023 Q1 and the operation was killed after the backup was completed. evidence below ⬇️ 

```shell=
2024-08-24 13:54:44,539 INFO clean_and_reparse.py::__backup:L49 :  Backup complete! Commencing clean and reparse.
Backup complete! Commencing clean and reparse.
Killed
```

![Screenshot 2024-08-24 100459](https://github.com/user-attachments/assets/8ec7de92-3cbe-4fae-a5ab-53d5a40bef0e)

I then tried another quarter: FY2023Q2 and couldn't proceed:

```shell=
vcap@fc474368-25ed-4bfd-51b7-c201:~$ python manage.py clean_and_reparse -y 2023 -q Q2

You have selected to reparse datafiles for FY 2023 and Q2. The reparsed files will NOT be stored in new indices and the old indices
These options will delete and reparse (20) datafiles.
Continue [y/n]? y
The latest ReparseMeta model's (ID: 2) timeout_at field is None. Cannot safely execute reparse, please fix manually.
```

Worth noting that FY23Q1 has a couple of large files that should generate a lot of errors, so I'd like to see how this operation performs before this in prod.

elipe17 committed:
- Updated docs

elipe17 committed:
- saving after each update

elipe17 commented:
> @elipe17 @andrew-jameson @jtimpe @raftmsohani I'm currently `blocked` on testing this PR in qasp environment. I attempted to reparse this morning for FY2023 Q1 and the operation was killed after the backup was completed. evidence below ⬇️
> 
> ```
> 2024-08-24 13:54:44,539 INFO clean_and_reparse.py::__backup:L49 :  Backup complete! Commencing clean and reparse.
> Backup complete! Commencing clean and reparse.
> Killed
> ```
> 
> ![Screenshot 2024-08-24 100459](https://private-user-images.githubusercontent.com/63075587/361146910-8ec7de92-3cbe-4fae-a5ab-53d5a40bef0e.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MjQ2Nzg3MjQsIm5iZiI6MTcyNDY3ODQyNCwicGF0aCI6Ii82MzA3NTU4Ny8zNjExNDY5MTAtOGVjN2RlOTItM2NiZS00ZmFlLWE1YWItNTNkNWE0MGJlZjBlLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDA4MjYlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwODI2VDEzMjAyNFomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTM0Y2FjZDZlNDJlZGI2YzZmOWRhYmRlNWM3OWU1MDIzZGY3NzQ4OWY1ZWU1NWU5YzVmZmU1YmQyMzFlNTQ1NDgmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.pAGFOphWrfuc8QZ2QLBJe9Vm0TlJEYZinsUhn17nVjk)
> 
> I then tried another quarter: FY2023Q2 and couldn't proceed:
> 
> ```
> vcap@fc474368-25ed-4bfd-51b7-c201:~$ python manage.py clean_and_reparse -y 2023 -q Q2
> 
> You have selected to reparse datafiles for FY 2023 and Q2. The reparsed files will NOT be stored in new indices and the old indices
> These options will delete and reparse (20) datafiles.
> Continue [y/n]? y
> The latest ReparseMeta model's (ID: 2) timeout_at field is None. Cannot safely execute reparse, please fix manually.
> ```
> 
> Worth noting that FY23Q1 has a couple of large files that should generate a lot of errors, so I'd like to see how this operation performs before this in prod.

@ADPennington I updated the meta model in qasp so that you can continue testing. The `Killed` console output indicates to me that the process was killed for some reason. I can't go far enough back in the logs to see if I can see exactly what happened.

elipe17 committed:
Merge branch 'develop' into 3064-reparse-meta

ADPennington commented:
@elipe17 latest test notes/questions below ⬇️ I didn't observe anything that needs to be addressed in this ticket; this is mostly for my SA. 

- Is there a way to know which source file id(s) are associated with the difference between the deleted/created counts? Looks like after the reparsing, the record count is different, which will sometimes be the case when validation is updated, but I imagine we'd also want to be able to investigate files to check if something went wrong? (see below):
![Screenshot 2024-08-28 083803](https://github.com/user-attachments/assets/e48118a8-4646-4517-8316-eb6e12679af1)

- what's the difference between total # of records initial and  # of records created? Is one capturing the number of records in files vs number of records in the db after reparsing?

- are we replacing the records in the db or adding new records? After reparsing FY23Q3, I see 6628 TANF T4s, and 3314 "new" TANF T4s. I was kind of expecting to see only "new" TANF T4s == "all" TANF T4s for this fiscal period. I'm assuming this is because more than one version of the FY23Q3 file was subject to reparsing? (see below). If true, this is another good justification for why we want to control which versions get reparsed (i.e. most recent :smile:) 
![Screenshot 2024-08-28 083002](https://github.com/user-attachments/assets/b66587fb-6886-486d-89bb-cafd401393c2)

- mentioned this async too, so this is just for reference, would be helpful for admins to know how to "fix manually" when observe logentries like the following: `The latest ReparseMeta model's (ID: 2) timeout_at field is None. Cannot safely execute reparse, please fix manually.`

elipe17 commented:
> @elipe17 latest test notes/questions below ⬇️ I didn't observe anything that needs to be addressed in this ticket; this is mostly for my SA.
> 
> * Is there a way to know which source file id(s) are associated with the difference between the deleted/created counts? Looks like after the reparsing, the record count is different, which will sometimes be the case when validation is updated, but I imagine we'd also want to be able to investigate files to check if something went wrong? (see below):
>   ![Screenshot 2024-08-28 083803](https://private-user-images.githubusercontent.com/63075587/362286211-e48118a8-4646-4517-8316-eb6e12679af1.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MjQ4NTAyNTUsIm5iZiI6MTcyNDg0OTk1NSwicGF0aCI6Ii82MzA3NTU4Ny8zNjIyODYyMTEtZTQ4MTE4YTgtNDY0Ni00NTE3LTgzMTYtZWI2ZTEyNjc5YWYxLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDA4MjglMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwODI4VDEyNTkxNVomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTk2NzZkNjRkZjIxNTc1N2M2YmIwNjM3ZTMzZThmMDAyZjZjYzk1OTA5NWJjNTg2NDI4OTU3M2UzMjYxZGZhZjImWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.bK25s8IU2NvoLwal5fkgSYd9EndL5yZy1RsrrVyZLaM)
> * what's the difference between total # of records initial and  # of records created? Is one capturing the number of records in files vs number of records in the db after reparsing?
> * are we replacing the records in the db or adding new records? After reparsing FY23Q3, I see 6628 TANF T4s, and 3314 "new" TANF T4s. I was kind of expecting to see only "new" TANF T4s == "all" TANF T4s for this fiscal period. I'm assuming this is because more than one version of the FY23Q3 file was subject to reparsing? (see below). If true, this is another good justification for why we want to control which versions get reparsed (i.e. most recent 😄)
>   ![Screenshot 2024-08-28 083002](https://private-user-images.githubusercontent.com/63075587/362286322-b66587fb-6886-486d-89bb-cafd401393c2.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MjQ4NTAyNTUsIm5iZiI6MTcyNDg0OTk1NSwicGF0aCI6Ii82MzA3NTU4Ny8zNjIyODYzMjItYjY2NTg3ZmItNjg4Ni00ODZkLTg5YmItY2FmZDQwMTM5M2MyLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDA4MjglMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwODI4VDEyNTkxNVomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTcyM2MzZWNjZGU4ODE5M2M5ZDQ4MzcwNDQ1MmY4NTc3ZmFiZjQxMzQ0ZDZjNWI2MzM3NDExYmI1NDhhYzJhOGMmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.anG5ePkn-QkpYg7Zy3MHslQzk46mgM67z80CYdr0QEY)
> * mentioned this async too, so this is just for reference, would be helpful for admins to know how to "fix manually" when observe logentries like the following: `The latest ReparseMeta model's (ID: 2) timeout_at field is None. Cannot safely execute reparse, please fix manually.`

@ADPennington, see my responses below :).

- The record count is/can be different for files that have not been cat4 validated. Since records with cat4 errors don't get serialized to the DB we can expect to see the "num records" fields to not always be a one-to-one match since cat4 is relatively new. We can write a spike ticket to investigate the feasibility of tracking before and after record counts for files, [this ticket](https://github.com/raft-tech/TANF-app/issues/2562) might be a way for us to get that information. As an intermediary, I have also written [this ticket](https://github.com/raft-tech/TANF-app/issues/3157) which adds some more useful fields. Specifically tracking cat4 errors before and after the reparse will help illuminate if it makes sense that the record counts have diverged.

- The two fields `Total num records initial` and `Total num records post` indicate the total number of records in the DB before and after the reparse event. The `Total num records deleted` field indicates how many records this reparse event deleted from the DB and the `Total num records created` indicates how many records were re-created during the reparse event.

- For reparsing, all records associated with the selected files are deleted from the DB and then recreated. No record duplication should be occuring.

- See the steps below to "fix manually"
1. Ssh into the appropriate backend
2. Enter the Django Shell Plus 
3. Execute:
```python
latest = ReparseMeta.get_latest()
```
4. Execute:
```python
latest.timeout_at = timezone.now()
```
5. Execute:
```python
latest.save()
```
6. You should now be able to execute a new reparse command.

ADPennington commented:
> * The record count is/can be different for files that have not been cat4 validated. Since records with cat4 errors don't get serialized to the DB we can expect to see the "num records" fields to not always be a one-to-one match since cat4 is relatively new. We can write a spike ticket to investigate the feasibility of tracking before and after record counts for files, [this ticket](https://github.com/raft-tech/TANF-app/issues/2562) might be a way for us to get that information. As an intermediary, I have also written [this ticket](https://github.com/raft-tech/TANF-app/issues/3157) which adds some more useful fields. Specifically tracking cat4 errors before and after the reparse will help illuminate if it makes sense that the record counts have diverged.

per async with @elipe17 #3096 is the ticket intended to capture more details about cat1 and cat4 errors in data file summaries. linking just for reference to related ideas 😄

ADPennington reviewed (approved):
LGTM 🚀
